### PR TITLE
Gorilla-style chunks

### PR DESCRIPTION
**Pull request by beorn7**
The culmination of all my preparations of the last weeks: Chunk encoding inspired by Facebook's Gorilla paper, joined with ideas from and adjustments to the Prometheus world.

On a real-world beefy Prometheus server, I have measured 1.77 bytes / sample (measured as bytes taken within the 1kB chunks). This is an upper bound as there were many uncompleted chunks. I'll keep reporting results as the server runs longer.

On disk (including all the indices, snapshot and everything), the same server currently uses 3.0 bytes / sample with 500k time series and 520M samples ingested so far. I expect this value to drop, too, as the server runs longer.

A self-monitoring Prometheus server needs less than 0.3 bytes / sample, which illustrates how far down we can go for special data set.

@juliusv @fabxc @brian-brazil Please have a look. Note that the existing tests (depsite not being very low-level) cover the code completely with the exception of error branches (like "unknown encoding"). The fuzz test I wrote ages ago was very useful in uncovering weird cornercases and bugs. Interestingly, almost none of the bugs were in the weird bit-twiddling code. It was mostly more high level (like reading an unsigned int where a signed had to be read or getting the iterator behavior wrong etc.).

@juliusv Please note that this PR depends on https://github.com/prometheus/prometheus/pull/1477 , which you have already 70% reviewed. I need your thumb for the final version. Another PR this depends upon is https://github.com/prometheus/prometheus/pull/1453 , too, which should be fairly easy to review. I (and be now the whole world) would be grateful if you could get to it. That PR has been created 13d ago.

@fabxc Please note that this PR depends on https://github.com/prometheus/prometheus/pull/1466 , which you have already 70% reviewed. I need your thumb for the final version. Another PR this depends upon is https://github.com/prometheus/prometheus/pull/1453 , too, which should be fairly easy to review. I (and be now the whole world) would be grateful if you could get to it. That PR has been created 13d ago.

### CODE REVIEW

File: storage/local/chunk.go [hunk 1]
```diff
@@ -306,6 +312,21 @@ func rangeValues(it chunkIterator, in metric.Interval) ([]model.SamplePair, erro
 	return result, it.err()
 }
 
+// addToOverflowChunk is a utility function that creates a new chunk as overflow
+// chunk, addse the provided sample to it, and returns a chunk slice containing
```
Comment (brian-brazil) [inline storage/local/chunk.go]: adds

**Comment by brian-brazil** [comment]
How does this compare CPU wise?

File: storage/local/gorilla.go [hunk 2]
```diff
@@ -0,0 +1,1103 @@
...
+
+// Gorilla chunk encoding is inspired by the following paper:
+//   Gorilla: A Fast, Scalable, In-Memory Time Series Database
+//   T. Pelkonen et al., Facebook Inc.
+//   http://www.vldb.org/pvldb/vol8/p1816-teller.pdf
+// Note that there are significant differences in detail, some of which due to
+// the way Prometheus chunks work, some of which to optimize for the Prometheus
+// use-case.
+//
+// Layout of a 1024 byte gorilla chunk (big endian, wherever it matters):
```
Comment (brian-brazil) [inline storage/local/gorilla.go]: Most users are going to be on little endian, shouldn't we save a tad of cpu by storing things that way?

File: storage/local/chunk.go (hunk 1 above)
Comment (beorn7) [inline storage/local/chunk.go]: Fixed.

File: storage/local/gorilla.go (hunk 2 above)
Comment (beorn7) [inline storage/local/gorilla.go]: Go doesn't work that way. You needed to use the `unsafe` package to ever interpret arbitrary byte sequences as a number. That would indeed be faster (but matter way more for our old chunks as they are all nicely byte aligned, while with Gorilla we need to bit shift all the time anyway, so the gain wouldn't be as large). 

If you look at the code in `encoding/binary`, you'll see that it doesn't make any use of the native endianess.

Using little endian would only make the code more difficult to read (even for Germans). Performance impact would be zero.

**Comment by beorn7** [comment]
CPU impact: I still have to do proper A/B testing. CPU usage depends on so many things. Preliminary estimates suggest the impact is not dramatic. Other stuff matters more (like the CPU savings because of my changes this PR is based upon are way larger).

The fuzz benchmark might be a good indicator. But note that it does random access of individual samples in the middle of a chunk all the time, which is rare in practice and more expensive for Gorilla-like chunks. Note that (as expected) there are fewer allocations (because you  need fewer chunks).

```
BenchmarkFuzzChunkType0-8              5         282778479 ns/op        20034462 B/op     412064 allocs/op
BenchmarkFuzzChunkType1-8              5         288646636 ns/op        20590161 B/op     425601 allocs/op
BenchmarkFuzzChunkType2-8              5         390896479 ns/op        14304241 B/op     291702 allocs/op
```

**Comment by beorn7** [comment]
And BTW, we are at 1.47 bytes / sample upper bound right now... Approaching the 1.37 from the Gorilla paper...

An disk, we are at 2.4 bytes / sample.

File: storage/local/gorilla.go [hunk 3]
```diff
@@ -0,0 +1,1103 @@
...
+	"math"
+
+	"github.com/prometheus/common/model"
+)
+
+// Gorilla chunk encoding is inspired by the following paper:
+//   Gorilla: A Fast, Scalable, In-Memory Time Series Database
+//   T. Pelkonen et al., Facebook Inc.
+//   http://www.vldb.org/pvldb/vol8/p1816-teller.pdf
+// Note that there are significant differences in detail, some of which due to
```
Comment (juliusv) [inline storage/local/gorilla.go]: s/some of which/some of which are/

or

s/some of which due/some due/

File: storage/local/gorilla.go [hunk 4]
```diff
@@ -0,0 +1,1103 @@
...
+func (c gorillaChunk) zeroDDTRepeats() (repeats uint64, offset uint16) {
+	offset = binary.BigEndian.Uint16(c[gorillaCountOffsetBitOffset:])
+	if offset == 0 {
+		return 0, 0
+	}
+	return c.readBitPattern(offset, 7) + 1, offset
+}
+
+func (c gorillaChunk) setZeroDDTRepeats(repeats uint64, offset uint16) {
+	if repeats == 0 {
```
Comment (fabxc) [inline storage/local/gorilla.go]: This should be a switch. Saves comparisons.

**Comment by fabxc** [comment]
Can you elaborate what one `op` is in the benchmark above?

**Comment by beorn7** [comment]
One op in the benchmark is one run through a random set of fuzzed samples. So quite a lot. But the random samples generated depend on a seed, so they are reproducible.

**Comment by juliusv** [comment]
Wow. Without having looked closer yet, it all sounds pretty awesome so far... thanks!

**Comment by fabxc** [comment]
So ~600-700 samples take 0.3ms.
That sounds like it can be heavily optimized – but I'm all for doing this at a later point.
Our read path is in trouble in other places right now.

**Comment by beorn7** [comment]
@fabxc Sorry, I was mistaken. The benchmark version of the fuzz test is different from the normal fuzz test. It uses in fact 100,000 samples per run. In more detail, it does the following, all in one goroutine:

1. Append 50,000 samples.
2. Read those 50,000 samples individually in random order.
3. Append another 50,000 samples.
4. Read all 100,000 samples individually in random order.

And that's what needs 400ms.

Now I'm inspired to do it concurrently with multiple goroutines.

**Comment by fabxc** [comment]
That seems way more realistic – I was a bit shocked.

**Comment by beorn7** [comment]
@fabxc I was talking about the fuzz benchmark. In reality, many retrievers (whatever they are called these days) call `Append` concurrently. But I was also half joking. I have more pressing issues than making the benchmark concurrent.

File: storage/local/gorilla.go (hunk 4 above)
Comment (beorn7) [inline storage/local/gorilla.go]: Done.

File: storage/local/gorilla.go (hunk 3 above)
Comment (beorn7) [inline storage/local/gorilla.go]: Done.

**Comment by beorn7** [comment]
For the record (pun intended): On the real-world test server, I'm now at 1.36 bytes/sample , i.e. now officially better than vanilla Gorilla. With total disk usage, we are at 1.9 bytes/sample, i.e. beating Influx's recently reported 2.2 bytes/sample, assuming they used a comparable metric here.

**Comment by beorn7** [comment]
I ran a proper A/B test now, with the same binary, just traditional double-delta chunks on the one, new Gorilla-style chunks on the other, with a real-life scrape config from within SC production. The servers ingest 15k samples/s and have a pretty expensive rule set to evaluate (many quantile calculation from histograms). Curious fact: Traditional double-delta is better than my older conservative estimate (3.3 instead of 4.5 bytes / sample). The CPU usage and rule evaluation time with Gorilla chunks has a periodic pattern, going down when a wave of new chunks is created, as accessing at the end of a chunk is more expensive.

Results:

Chunk type | bytes/sample | cores | evaluation cycle
------|-----:|----:|----:
double-delta | 3.3 | 1.6 | 2.9s
Gorilla | 1.3 | 2.4 | 4.9s

**Comment by beorn7** [comment]
I also found a rarely tickled iterator bug during the tests, which I still have to hunt down.

**Comment by brian-brazil** [comment]
0.5x the space for 1.5x the cpu doesn't sound like a bad tradeoff. The increase in evaluation time could be more of a problem.

**Comment by beorn7** [comment]
It's 0.39x the space for 1.5x the CPU. Evaluation time will depend on circumstances. But the user will anyway be able to pick their favorite chunk. We even kept single-delta chunks around for that reason. Now there is a more meaningful trade-off.

**Comment by fabxc** [comment]
All great so far. Just one last thing.

As mentioned in person before, this should not be named Gorilla encoding.

- this encoding has semantic differences from the one of the Gorilla paper.
- the fundamentals of this XOR encoding were not invented by Facebook. If we want to attribute by our naming, we would have to go back at least one step further. Even then...
- it is not on us to name the encoding someone else presented. If we want to name our variation, that's fine – but just taking the name of Facebook's internal system, when they left it nameless themselves seems wrong.

Attribution to what this was inspired by should be absolutely clear, of course. Ideally with references to the Gorilla paper and the references this one provides regarding the compression format.

I vote to keep the name neutral ("xor" – in the spirit of "delta" and "double-delta") or – if necessary – something original.

**Comment by beorn7** [comment]
Happy to hear suggestions for names. I like "Gorilla-style" as an internal name because everybody involved in the development knows immediately what is meant. And the "...-style" suggests that it's not the same. In the "public" interface, i.e. documentation and output of `prometheus -h`, we could make it clearer as right now (it's called "Gorilla-style" there). The question is if we really want to change code internals like file names and identifiers within the code. If we have a short name that tells the developer equally clear as "gorilla" what we are dealing with, fine with me.

About "xor": That would be misleading as it is only one of the value encodings and never used for timestamp encoding. Obviously, as vanilla Gorilla uses xor for all values (while I'd assume in practice that less than 50% of chunks will have xor value encoding), that would be an argument against Gorilla as a name, too.

**Comment by beorn7** [comment]
I'm now running the beefy "real life" test setup with the latest state of things.
If that doesn't expose further problems, I'd say we should merge this (modulo name change, if required), but leave type-1 chunks as default, so only users who actively want type-2 will benefit (and suffer) from them.

**Comment by beorn7** [comment]
Test run looks good so far.

Any ideas about the naming?

I was thinking along "variable bit width" -> vbw. But that's horribly un-catchy...

**Comment by fabxc** [comment]
I'm don't find catchyness necessary.

The issue with "internal naming" is, that it doesn't exist. "Internally, we
call these Gorilla-chunks" will be dropped in every talk and conversation.
Then we constantly have to correct wrong assumptions caused by that.

**Comment by beorn7** [comment]
So should I go with "vbw" then?

**Comment by beorn7** [comment]
Test server still runs without problem.

About the renaming: After talking to @juliusv and @fabxc in person, we settled on "varbit" as a somewhat meaningful and still catchy name. I'm performing the renaming now.

**Comment by beorn7** [comment]
OK, done.

Need a thumb for merging. ;)

**Comment by fabxc** [comment]
:+1: enable compression over-drive :rocket:

**Pull request merged**